### PR TITLE
Source Jira: Add pull requests stream

### PR DESCRIPTION
**Pull request by cjwooo**
## What
Adds a Pull Requests stream to fetch data on linked pull requests for Jira issues. Currently only supports GitHub pull requests.
Also fixes the field_ids_by_name() method on the IssueFields stream to handle multiple field ids per field name, a case we've seen with some of our projects.

## How
Uses an undocumented Jira API to fetch data on linked pull requests. Only runs on Jira issues that have been determined to have linked pull requests via their "Development" field.

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 1]
```diff
@@ -770,6 +776,62 @@ def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwarg
...
+class PullRequests(IncrementalJiraStream):
+    """
+    Undocumented internal API used by Jira webapp
+    """
+
+    cursor_field = "updated"
+    parse_response_root = "detail"
+    # raise_on_http_errors = False
+
+    pr_regex = r"(?P<prDetails>PullRequestOverallDetails{openCount=(?P<open>[0-9]+), mergedCount=(?P<merged>[0-9]+), declinedCount=(?P<declined>[0-9]+)})|(?P<pr>pullrequest={dataType=pullrequest, state=(?P<state>[a-zA-Z]+), stateCount=(?P<count>[0-9]+)})"
```
[the remark this answers is not in the document]
Comment (cjwooo) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py:788]: We use this to associate closed Jira issues with the Github pull requests that resolved the issue.

[the remark this answers is not in the document]
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py:788]: @sherifnada can you give your opinion here?

**Comment by sherifnada** [comment]
@cjwooo thanks for opening this PR! I took a look around online and it while this endpoints seems(?) to maybe have held up via [Hyrum's law](https://www.hyrumslaw.com/), it's difficult to support adding it to the connector because it's undocumented; if it changes or breaks or gets removed at any point, it'll cause user disruption. We could always say something like "well, blame JIRA because they removed it" but the problem will likely take the shape of undocumented/unannounced changes which cause ongoing maintenance work on this connector. How have you been thinking about this issue in the context of your product?

**Comment by cjwooo** [comment]
@sherifnada If this API breaks suddenly, we currently inform the user, via sync logs, that we can't fetch linked pull requests, and continue pulling the rest of our equivalent of Airbyte streams normally. I can implement a similar strategy here, where the Pull Requests stream will log that the stream is broken due to changes in the Jira API, without failing the entire source sync.

**Comment by cjwooo** [comment]
@sherifnada Would that be acceptable?

**Comment by marcosmarxm** [comment]
(waiting updates from user adding safeguards)

**Comment by cjwooo** [comment]
@marcosmarxm This is ready for review.

**Comment by marcosmarxm** [comment]
Thanks @cjwooo

**Comment by cjwooo** [comment]
Bump

**Comment by marcosmarxm** [comment]
@cjwooo we're waiting the solution for https://github.com/airbytehq/airbyte/issues/9017 to merge your publish your contribution. Sorry the long delay.

**Review by marcosmarxm — APPROVED** [review]
sorry the long delay @cjwooo

**Comment by marcosmarxm** [comment]
@cjwooo could you give me permission to your repo? Or solve the conflicts so I can merge your contribution? If helps please see my PR: https://github.com/airbytehq/airbyte/pull/8559

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 2]
```diff
@@ -857,6 +863,68 @@ def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwarg
...
+        super().__init__(**kwargs)
+        self.issues_stream = issues_stream
+        self.issue_fields_stream = issue_fields_stream
+
+    @property
+    def url_base(self) -> str:
+        return f"https://{self._domain}/rest/dev-status/1.0/"
+
+    def path(self, **kwargs) -> str:
+        return f"issue/detail"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:891]:
```suggestion
        return "issue/detail"
```
remove f-string is breaking the format

**Comment by marcosmarxm** [comment]
The connector was published in https://github.com/airbytehq/airbyte/pull/8559 I'm closing this PR and merging the other to relese the new version of the connector.

**Pull request closed without merging**